### ISSUE TITLE
diag(p3): hidden_size sweep for BC-fit (Phase 1.5 escalation if #278 fails) — test capacity hypothesis

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#280]
## Hypothesis being tested

If class-balanced BC (#279) still fails the WORK-house discriminating sub-task, then the failure is **capacity-bound**, not loss-distribution-bound. This experiment tests whether scaling `hidden_size` (current default 64, ~8.1k params) fixes the representational gap.

## Context (Curator update — 2026-05-17)

PR #278's BC-init builder produced a **cross-finding** that may obviate the original 4-cell sweep:

- 10k pairs / 10 epochs @ `hidden_size=64` → `gap_closed = 0.368`, house-on-WORK = 31.5% — reproduces #272's failure mode
- 40k pairs / 30 epochs @ `hidden_size=64` → `gap_closed = 0.934` — **BC succeeded at the default size**

This suggests the architecture has sufficient capacity at default `hidden_size=64`; the original "INDUCTIVE_BIAS_GAP" verdict from #272 (PR #276) is more likely a training-budget artifact than a representational ceiling. Builder has discretion to scope this experiment down (see "Recommended scope" below).

## Recommended scope (after #278 cross-finding)

Two paths, in increasing cost order. **Pick the smallest path that resolves the open question for the cell where #279 (class-balanced BC) still falls short.**

### Path A (minimal — recommended default): 2-cell capacity probe at short-budget regime

Goal: confirm capacity is not the bottleneck when budget is short. Compare at the low-budget setting where the failure originally appeared.

- `hidden_size ∈ {64, 128}` at **10k pairs / 10 epochs** (the budget where #272/PR #276 saw failure)
- If 128 lifts house-on-WORK significantly while 64 stays at chance → capacity matters at short budget (file: increase default and/or use longer budget in production)
- If both stay at chance → capacity isn't the limiting factor at short budget; class imbalance or data quantity dominates → defer to #279/#281 findings

### Path B (full sweep — only if Path A is inconclusive or #279 verdict 3 demands it)

- `hidden_size ∈ {64, 128, 256, 512}` at 10 epochs (extend to 30 if loss still decreasing — PR #276 noted loss was actively decreasing at epoch 10, *underfitting* not overfitting)
- Use class-balanced loss if #279's verdict indicated partial improvement (see "What BC loss to track" below)

For each cell, report: final eval loss, house-on-WORK accuracy, joint accuracy, parameter count.

## What BC loss to track

Use `bc_fit_only.py`'s existing per-head accuracy breakdown (see `compute_verdict`):

- `final.eval_loss` — sum-of-CE over the three heads `[house, mode, signal]`
- `final.house_acc` — house head, all rows (dominated by REST class prior P(REST)≈0.9435)
- `final.mode_acc` — REST/WORK 2-way
- `final.signal_acc` — 2-way
- `final.joint_acc` — exact match across all 3 heads
- **`result.house_acc_on_work_subset`** — the discriminating metric. This is the eval-set accuracy of the house head restricted to rows where the specialist labeled mode=WORK. Chance level for 3 owned houses is ≈0.354. A successful fit must reach ≥0.90 here. **This is the headline number for verdict logic.**
- `result.n_params` — for capacity-vs-accuracy plot

## Verdict logic (capacity-vs-data interpretation)

| house-on-WORK pattern across cells | Interpretation | Next step |
|---|---|---|
| **64 already saturates (≥0.90)** at short budget | Capacity wasn't the gap. If #279 also passed → resolved by class-balancing. If #279 failed → reframe research direction; specialist may use info not in obs. | Close ladder. File ppo-help-brief update. |
| **Monotonic improvement with hidden_size**, plateau ≥0.90 by 128 or 256 | Confirmed capacity bound at short budget. PPO's default-64 actor is undersized for this discriminating sub-task. | File production-PPO follow-up with winning `hidden_size` (likely 128 or 256) |
| **Plateau below 0.90 even at 512** | Architecture mismatch isn't pure capacity — it's *kind* of capacity (depth? attention? non-linearity?). | Promote #281 (TransformerPolicyNetwork). |
| **No cell exceeds chance (~0.35)** | Neither capacity nor class-balance #279 is the gap. Information for the specialist's argmax over burning-owned houses may not be in the observation. | Reframe: inspect what `obs[i]` actually contains; specialist may use privileged info. |

## Implementation guidance

`experiments/p3_specialization/bc_fit_only.py` already accepts `--hidden-size` (default 64) and emits the metrics above. The sweep is a thin bash wrapper:

```bash
for HS in 64 128; do  # or 64 128 256 512 for Path B
  uv run python experiments/p3_specialization/bc_fit_only.py \
    --hidden-size "$HS" \
    --num-steps 2500 --epochs 10 \
    --out "experiments/p3_specialization/runs/280_hs${HS}.json"
done
```

If #279's verdict says class-balancing helped partially: add a `--class-balanced` flag to `bc_fit_only.py` (mirror whatever #279 ships) and pass it for all cells. Otherwise run the existing code as-is.

Aggregate the JSON outputs and report the table in the issue comment / notebook.

## Affected Files

- `experiments/p3_specialization/bc_fit_only.py` — existing; runner accepts `--hidden-size`. No changes expected unless adding `--class-balanced` per #279's outcome (see `bc_fit_only.py:287` for arg, `bc_fit_only.py:136-178` for `train_bc`).
- `bucket_brigade/training/networks.py` — reference only. `PolicyNetwork` definition at lines 13-99; `hidden_size: int = 64` default at line 45.
- `experiments/p3_specialization/runs/280_hs{64,128,...}.json` — new per-cell output files.
- `experiments/p3_specialization/analyze_280.py` (new, optional) — aggregator that reads the per-cell JSONs and emits the verdict per the matrix above. Pattern: see `experiments/p3_specialization/analyze_270.py`.
- Project memory: `~/.claude/projects/-Users-rwalters-GitHub-bucket-brigade/memory/project_ppo_failure_mode.md` — update with verdict outcome.

## Test Plan

- [ ] Manual verification: run Path A (2 cells). Confirm each cell emits a valid result JSON with `result.house_acc_on_work_subset` present and finite.
- [ ] Automated tests: N/A (this is a diagnostic experiment, not production code). If a `--class-balanced` flag is added to `bc_fit_only.py`, add a smoke test alongside whatever #279 ships.
- [ ] Verdict applied: produce a markdown table of `{hidden_size, n_params, eval_loss, house_acc, house_acc_on_work_subset, joint_acc}` for each cell and apply the verdict from the matrix above.
- [ ] Cross-reference: if winning `hidden_size > 64`, file a production-PPO follow-up issue that uses the winning size in `experiments/p3_specialization/train.py`. If `64` saturates and #279 also passed, comment on this issue with the resolution and link to the relevant project-memory update.
- [ ] Edge case: confirm that `hidden_size=512` PolicyNetwork still loads and trains under the existing trainer (no shape mismatch with PPO's downstream consumers if production follow-up is filed).

## Dependencies

- [ ] #279: Class-balanced BC verdict (must complete first — determines whether to use `--class-balanced` flag in this sweep, and whether this experiment is needed at all)

If #279 shows class-balancing already lifts house-on-WORK to ≥0.90 at `hidden_size=64`, this issue can be closed as resolved without running cells.

Note: PR #278's BC-init cross-finding (40k/30ep at hidden_size=64 → gap_closed 0.934) is *advisory context*, not a strict dependency — it suggests but does not prove that #280 is unnecessary at full training budget.

## Cost

- Path A: ~30-60 min total (2 cells × ~15-30 min each, pure CPU).
- Path B: ~2-3 hours total (4 cells × ~30 min each, pure CPU).

Run on `COMPUTE_HOST_PRIMARY` per CLAUDE.md's compute guidelines. Do not run locally.

## References

- PR #276 (the unbalanced #272 baseline this builds on)
- PR #278 (BC-init builder's cross-finding: BC succeeded at default `hidden_size=64` with 40k/30ep)
- #279 (class-balanced BC — must complete first; verdict determines whether to use class-balancing here)
- #281 (TransformerPolicyNetwork escalation — gated on this issue's verdict)
- `bucket_brigade/training/networks.py:13-99` (PolicyNetwork; default `hidden_size=64`)
- `experiments/p3_specialization/bc_fit_only.py:275-298` (CLI arg parsing)
- `docs/ppo-help-brief.md`

**Comment by rjwalters** [comment]
## Cross-issue finding from PR #278 (BC-init builder): hidden_size sweep may be unnecessary

While implementing #270's BC-init pipeline, PR #278's builder did an ablation comparing BC training budgets at the **default `hidden_size=64`**:

- **10k pairs / 10 epochs** (same as PR #276's setup that motivated this issue): `gap_closed = 0.368`, house-on-WORK = 31.5% — **matches the failure mode #272 reported**
- **40k pairs / 30 epochs (same `hidden_size=64`)**: `gap_closed = 0.934` — **BC succeeded**

**This suggests the architecture has sufficient capacity at the default size.** #272's "INDUCTIVE_BIAS_GAP" verdict is more likely a training-budget artifact than a representational ceiling.

## Recommendation for whoever builds this

Before running the full 4-cell hidden_size sweep, consider:
- Does the 0.934 result at 40k/30-epoch already settle the capacity question?
- If yes, this issue can be closed as "resolved by PR #278's cross-finding"
- If you still want to confirm, run just `hidden_size=64` vs `hidden_size=128` at the smaller (10k/10-epoch) budget to test if capacity helps where budget is short

You have full discretion to scope this down based on what's already established. The original 4-cell sweep is no longer the most efficient experiment.

## References
- PR #278 (the cross-finding source)
- PR #276 (#272 verdict that originally motivated this issue)
- Judge of #276 — class-imbalance hypothesis the cross-finding effectively confirms

**Comment by rjwalters** [comment]
Curating this issue per user request

**Comment by rjwalters** [comment]
Curation complete (2026-05-17).

**What changed**:
- Added Curator context summarizing PR #278's cross-finding (BC succeeded at default `hidden_size=64` with 40k/30ep budget — capacity may not be the gap).
- Recommended scope reduction: Path A (2-cell probe at `hidden_size ∈ {64, 128}`, short-budget regime) as the default; Path B (full 4-cell sweep) only if Path A is inconclusive or #279 demands it.
- Specified the BC metrics to track, with `result.house_acc_on_work_subset` as the headline number (chance ≈0.354 for 3 owned houses; success threshold ≥0.90).
- Refined capacity-vs-data interpretation matrix into 4 verdict rows.
- Added Implementation Guidance with a concrete bash wrapper invocation.
- Added Affected Files section citing `networks.py:13-99` and `bc_fit_only.py:275-298`.
- Added Test Plan section with manual verification, verdict-application, and edge-case checks.
- Formalized the Dependencies section gating on #279; applied `loom:blocked`.

**Why blocked**: #279 (class-balanced BC) is the strict gate — its verdict determines whether a `--class-balanced` flag is needed in this sweep, and may obviate the experiment entirely. PR #278's cross-finding is *advisory*, not a strict dependency.

**Duplicates**: Checked — distinct from #279 (class-balancing axis) and #281 (architecture-kind axis). All three target different rungs of the Phase 1.5 BC-fit ladder.

Awaiting human approval (`loom:issue`) once #279 lands and the conditional path is determined.
